handler/src/main/java/io/netty/handler/ssl/TlsMasterKeyHandler.java

```diff
@@ -0,0 +1,186 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.internal.tcnative.SSL;
+import io.netty.util.internal.ReflectionUtil;
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLSession;
+import java.lang.reflect.Field;
+
+/**
+ * The {@link TlsMasterKeyHandler} is a channel-handler you can include in your pipeline to consume the master key
+ * & session identifier for a TLS session.
+ * This can be very useful, for instance the {@link WiresharkTlsMasterKeyHandler} implementation will
+ * log the secret & identifier in a format that is consumable by Wireshark -- allowing easy decryption of pcap/tcpdumps.
+ */
+public abstract class TlsMasterKeyHandler extends ChannelInboundHandlerAdapter {
+
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(TlsMasterKeyHandler.class);
+
+    /**
+     * The JRE SSLSessionImpl cannot be imported
+     */
+    private static final Class<?> SSL_SESSIONIMPL_CLASS;
+
+    /**
+     * The master key field in the SSLSessionImpl
+     */
+    private static final Field SSL_SESSIONIMPL_MASTER_SECRET_FIELD;
+
+    /**
+     * A system property that can be used to turn on/off the {@link TlsMasterKeyHandler} dynamically without having
+     * to edit your pipeline.
+     * <code>-Dio.netty.tls.masterKeyHandler=true</code>
+     */
+    public static final String SYSTEM_PROP_KEY = "io.netty.tls.masterKeyHandler";
+
+    /**
+     * The unavailability clause of whether the private Sun implementation of SSLSessionImpl is available.
+     */
+    private static final Throwable UNAVAILABILITY_CAUSE;
+
+    static {
+        Throwable cause = null;
+        Class<?> clazz = null;
+        Field field = null;
+        try {
+            clazz = Class.forName("sun.security.ssl.SSLSessionImpl");
+            field = clazz.getDeclaredField("masterSecret");
+            cause = ReflectionUtil.trySetAccessible(field, true);
+        } catch (Throwable e) {
+            cause = e;
+            logger.debug("sun.security.ssl.SSLSessionImpl is unavailable.", e);
+        }
+        UNAVAILABILITY_CAUSE = cause;
+        SSL_SESSIONIMPL_CLASS = clazz;
+        SSL_SESSIONIMPL_MASTER_SECRET_FIELD = field;
+    }
+
+    /**
+     * Constructor.
+    */
+    public TlsMasterKeyHandler() {
```

Comment:
as this class is abstract this should be protected

Reply:
done.